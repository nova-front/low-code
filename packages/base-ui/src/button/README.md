# React Headless 按钮组件

## 功能需求分析

这个无样式按钮组件应该具备：

✅ 基础按钮功能（点击事件）
✅ 完善的类型检查（TypeScript）
✅ 完整的HTML按钮属性支持
✅ 无障碍（a11y）支持
✅ 禁用状态处理
✅ 组件类型扩展能力
✅ 不包含任何内联样式
✅ 支持通过className添加自定义样式
✅ 支持React Ref直接访问DOM元素
✅ 支持渲染为其他元素（如a标签）

## 关键特性实现说明

### ​无样式设计​​:

- 组件不包含任何内联样式或默认CSS
- 用户可以通过className属性添加自己的样式
- 默认添加unstyled-button类名便于样式定位

### ​类型安全​​：

- 使用TypeScript定义了完整的props类型
- 支持所有标准的按钮属性和链接属性

### ​​无障碍支持​​：

- 禁用状态时设置aria-disabled属性
- 作为链接使用时添加role="button"
- 禁用状态下阻止键盘焦点导航

### ​功能灵活性​​：

- 支持通过as属性渲染为不同元素类型
- 禁用状态下阻止所有交互事件
- 使用forwardRef支持React Ref访问

### ​完整API支持​​：

- 兼容所有标准按钮属性(onClick, type等)
- 支持所有标准链接属性(href, target等)
- 正确处理disabled和href属性的冲突

### ​开发体验优化​​：

- 添加了`displayName`便于调试
- 严格的事件处理程序类型检查
- 禁用状态下的完整事件阻断

## 优势总结

这个无样式按钮组件提供了:

- ​最大灵活性​​：完全由用户控制样式
- ​高兼容性​​：支持所有按钮/链接标准属性
- 专业级实现​​：考虑无障碍、禁用状态等专业需求
- 易于扩展​​：通过as属性支持多种渲染场景
- 开发友好​​：完整的TypeScript类型支持

该实现专注于提供可靠的核心功能，同时将样式控制权完全交给开发者，非常适合作为组件库的基础按钮组件。
您可以直接导入使用这个Button组件，通过className添加自定义样式，也可以将其作为基础构建更复杂的按钮组件。
