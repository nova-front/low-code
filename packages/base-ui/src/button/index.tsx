import React, {
  forwardRef,
  ButtonHTMLAttributes,
  AnchorHTMLAttributes,
  useMemo,
  useCallback,
} from 'react';
import { createPrefixedClassName, getA11yProps } from '../utils';

// 基础按钮属性定义
interface BaseButtonProps {
  children?: React.ReactNode;
  className?: string;
  disabled?: boolean;
  as?: React.ElementType;
}

// 简化的按钮属性类型
type ButtonProps = BaseButtonProps &
  Omit<ButtonHTMLAttributes<HTMLButtonElement>, keyof BaseButtonProps> &
  Omit<AnchorHTMLAttributes<HTMLAnchorElement>, keyof BaseButtonProps>;

// 工具函数：获取元素属性
const getElementProps = (
  element: React.ElementType,
  disabled: boolean,
  restProps: Record<string, any>
): Record<string, any> => {
  const elementProps: Record<string, any> = {};

  if (element === 'button') {
    elementProps.type = restProps.type || 'button';
    elementProps.disabled = disabled;
  } else {
    // 对于非按钮元素，使用工具函数获取可访问性属性
    const a11yProps = getA11yProps(disabled, 'button');
    // 使用展开运算符代替 Object.assign
    Object.keys(a11yProps).forEach((key) => {
      elementProps[key] = a11yProps[key];
    });

    // 特殊处理：如果是链接元素，需要移除 role
    if (element === 'a') {
      delete elementProps.role;
    }
  }

  return elementProps;
};

// 工具函数：过滤属性
const filterProps = (
  element: React.ElementType,
  props: Record<string, any>
): Record<string, any> => {
  const filteredProps = { ...props };

  if (element !== 'button') {
    delete filteredProps.type;
  }

  return filteredProps;
};

export const Button = forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const {
    children,
    className = '',
    disabled = false,
    as: Element = 'button',
    onClick,
    ...restProps
  } = props;

  // 使用 useMemo 优化类名计算
  const buttonClasses = useMemo(
    () => createPrefixedClassName('unstyled-button', className, { disabled }),
    [className, disabled]
  );

  // 使用 useMemo 优化元素属性计算
  const elementProps = useMemo(
    () => getElementProps(Element, disabled, restProps),
    [Element, disabled, restProps]
  );

  // 使用 useMemo 优化属性过滤
  const filteredProps = useMemo(
    () => filterProps(Element, restProps),
    [Element, restProps]
  );

  // 使用 useCallback 优化点击事件处理
  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      if (disabled) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      // 使用类型断言来处理事件类型
      if (onClick) {
        onClick(event as any);
      }
    },
    [disabled, onClick]
  );

  // 使用 useCallback 优化键盘事件处理
  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLElement>) => {
      if (disabled) {
        return;
      }

      // 对于非按钮元素，支持 Enter 和 Space 键触发点击
      if (
        Element !== 'button' &&
        (event.key === 'Enter' || event.key === ' ')
      ) {
        event.preventDefault();
        handleClick(event as any);
      }

      // 调用原始的 onKeyDown 处理器
      if (restProps.onKeyDown) {
        (restProps.onKeyDown as any)(event);
      }
    },
    [disabled, Element, handleClick, restProps.onKeyDown]
  );

  return (
    <Element
      {...filteredProps}
      ref={ref}
      className={buttonClasses}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      {...elementProps}
    >
      {children}
    </Element>
  );
});

Button.displayName = 'UnstyledButton';
