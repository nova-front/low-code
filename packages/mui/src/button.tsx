"use client";

import { ReactNode } from "react";
import ButtonBox from "@mui/material/Button";

interface ButtonProps {
  children: ReactNode;
  className?: string;
  appName: string;
}

export const Button = ({ children, className, appName }: ButtonProps) => {
  return (
    <div>
      <ButtonBox
        className={className}
        variant="contained"
        onClick={() => alert(`Hello from your ${appName} app!`)}
      >
        {children}
      </ButtonBox>
    </div>
  );
};
