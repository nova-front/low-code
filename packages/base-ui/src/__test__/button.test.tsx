import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Button } from "../button"; // 根据实际路径调整

describe("Button 组件", () => {
  // 测试1: 渲染基础内容
  it("正确渲染子内容", () => {
    render(<Button appName="Test App">Click Me</Button>);
    expect(screen.getByText("Click Me")).toBeInTheDocument();
  });

  // 测试2: 应用自定义 className
  it("接受并应用 className", () => {
    render(
      <Button appName="Test App" className="custom-class">
        Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("custom-class");
  });

  // 测试3: 点击事件触发 alert
  it("点击时触发带 appName 的 alert", () => {
    // 模拟 window.alert
    window.alert = jest.fn();
    const appName = "My Awesome App";

    render(<Button appName={appName}>Test</Button>);
    fireEvent.click(screen.getByRole("button"));

    expect(window.alert).toHaveBeenCalledTimes(1);
    expect(window.alert).toHaveBeenCalledWith(
      `Hello from your ${appName} app!`
    );
  });

  // 测试4: 必需的 appName prop
  it("需要 appName prop", () => {
    // 故意不传 appName 来测试 TypeScript 类型检查
    // 这里主要验证 TypeScript 的行为，实际运行时测试可能不需要
    const button = <Button appName="Test">Child</Button>;
    expect(button).toBeTruthy();
  });

  it("渲染快照匹配", () => {
    const { asFragment } = render(
      <Button appName="Snapshot Test">Snapshot</Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("具有正确的按钮角色", () => {
    render(<Button appName="A11y Test">Accessible</Button>);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("处理空子内容", () => {
    render(<Button appName="Empty Test">{""}</Button>);
    expect(screen.getByRole("button")).toBeEmptyDOMElement();
  });
});
