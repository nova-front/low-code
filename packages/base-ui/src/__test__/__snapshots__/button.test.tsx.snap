// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 组件 渲染快照匹配 1`] = `
<DocumentFragment>
  <button>
    Snapshot
  </button>
</DocumentFragment>
`;
